quil-plotting/quil_plotting/dataframe.py

```diff
@@ -0,0 +1,346 @@
+            "Gate": "category",
+            "Qubit": "category",
+            "Pulse": "category",
+            "Waveform": "category",
+            "Start Time (s)": float,
+            "End Time (s)": float,
+            "Duration (s)": float,
+            "Pulse Time (s)": float,
+            "Frame": "category",
+            "Channel Type": "category",
+            "IQ": complex,
+            "Offset": float,
+            "Color": "category",
+            "Color Faded": "category",
+            "Normalized IQ": complex,
+            "Label": "category",
+            "Runner": "category",
+        }
+    )
+
+    return df
+
+
+def deoxidize(num) -> complex:
+    """Remove Rust from a Number or PrefixExpression."""
+    num = num.inner()
+    if isinstance(num, PrefixExpression):
+        if num.operator == PrefixOperator.Plus:
+            return num.expression.evaluate({}, {})
+        elif num.operator == PrefixOperator.Minus:
+            return -1 * num.expression.evaluate({}, {})
+        else:
+            raise ValueError()
+    return num
+
+
+def get_iqs(pulse: Pulse, waveforms: Dict, frame):
+    waveform_reference = pulse.waveform
+    name = waveform_reference.name
+    parameters = pulse.waveform.parameters
+    sample_rate = frame["SAMPLE-RATE"].inner().to_real()
+    match name:
+        case "flat":
+            waveform_envelope = FlatWaveform(**parameters)
+            iqs = np.asarray(waveform_envelope.samples(sample_rate))
+        case "drag_gaussian":
+            waveform_envelope = DragGaussianWaveform(**parameters)
+            iqs = np.asarray(waveform_envelope.samples(sample_rate))
+        case "gaussian":
+            waveform_envelope = GaussianWaveform(**parameters)
+            iqs = np.asarray(waveform_envelope.samples(sample_rate))
+        case "erf_square":
+            waveform_envelope = ErfSquareWaveform(**parameters)
+            iqs = np.asarray(waveform_envelope.samples(sample_rate))
+        case _:
+            if name in waveforms:
+                iqs = np.array([deoxidize(val) for val in waveforms[name].matrix])
+            else:
+                raise ValueError(f"Waveform {name} not in waveforms")
+
+    ts = np.arange(0, len(iqs)) / sample_rate
+    return ts, iqs
+
+
+def program_to_dataframe(
+    program: Program,
+) -> pd.DataFrame:
+    """Expand the schedule into a dataframe of IQ values and metadata.
+
+    :param program:
+    """
+    # Get the instructions and the schedule
+    if hasattr(program, "_program"):
+        quil_program = program._program
+    elif isinstance(program, Program):
+        quil_program = program
+    else:
+        raise ValueError("Program must be a pyquil.Program or a quil.Program")
+
+    calibrations = quil_program.calibrations
+    expanded_quil_program, source_map = expand_with_source_mapping(quil_program, include_pragmas=False)
+    blocks = expanded_quil_program.control_flow_graph().basic_blocks()
+    assert len(blocks) == 1
```

Comment:
`ruff` should raise a lint for this. In short, `assert` is removed from optimized Python code, and it's generally better to explicitly raise an error. [Reference](https://docs.astral.sh/ruff/rules/assert/)